src/MainMenu.cpp

```diff
@@ -311,3 +316,235 @@ ZunResult MainMenu::DrawStartMenu(void)
+                            // FIXME: wrong assembly
+                            memcpy(&this->replayFileData[replayFileIdx], replayData, 0x50);
+                            sprintf(this->replayFilePaths[replayFileIdx], "./replay/%s", replayFileInfo.cFileName);
+                            sprintf(this->replayFileName[replayFileIdx], "User ");
+                            replayFileIdx++;
+                        }
+                        free(replayData);
+                        if (!FindNextFileA(replayFileHandle, &replayFileInfo))
+                            break;
+                    }
+                }
+                FindClose(replayFileHandle);
+                _chdir("../");
+                this->replayFilesNum = replayFileIdx;
+                this->unk_81fc = 0;
+                this->wasActive = this->isActive;
+                this->isActive = 0;
+                this->gameState = STATE_REPLAY_ANIM;
+                anmVm = this->vm;
+                for (cur = 0; cur < ARRAY_SIZE_SIGNED(this->vm); cur++, anmVm++)
+                {
+                    anmVm->pendingInterrupt = 15;
+                }
+                this->cursor = 0;
+            }
+            break;
+        }
+        break;
+    case STATE_REPLAY_UNLOAD:
+        if (this->stateTimer == 0x24)
+        {
+            this->gameState = STATE_STARTUP;
+            this->stateTimer = 0;
+        }
+        break;
+    case STATE_REPLAY_ANIM:
+        if (this->stateTimer < 0x28)
+        {
+            break;
+        }
+        if (this->replayFilesNum != NULL)
+        {
+            MoveCursor(this, this->replayFilesNum);
+            this->chosenReplay = this->cursor;
+            if (WAS_PRESSED(KEY_SHOOT | KEY_ENTER))
+            {
+                this->gameState = STATE_REPLAY_SELECT;
+                anmVm = &(this->vm[97]);
+                for (cur = 0; cur < 0x19; cur += 1, anmVm++)
+                {
+                    anmVm->pendingInterrupt = 0x11;
+                }
+                anmVm = &this->vm[99 + this->chosenReplay];
+                anmVm->pendingInterrupt = 0x10;
+                this->stateTimer = 0;
+                this->cursor = 0;
+                g_SoundPlayer.PlaySoundByIdx(10, 0);
+                this->currentReplay = (ReplayData *)FileSystem::OpenPath(this->replayFilePaths[this->chosenReplay], 1);
+                ValidateReplayData(this->currentReplay, g_LastFileSize);
+                for (cur = 0; cur < ARRAY_SIZE_SIGNED(this->currentReplay->stageScore); cur++)
+                {
+                    if (this->currentReplay->stageScore[cur] != NULL)
+                    {
+                        this->currentReplay->stageScore[cur] =
+                            (StageReplayData *)((u32)this->currentReplay + (u32)this->currentReplay->stageScore[cur]);
+                    }
+                }
+
+                do
+                {
+                    // FIXME: there's an additional jump
```

Comment:
My gut feeling is that the whole do/while/goto/return is slightly wrong somehow, but I'm having a hard time figuring out what the proper control flow should be. My gut feeling is that the `goto` should be a break, and the return might be gated behind an if or something?